config/load.go

```diff
@@ -0,0 +1,298 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/adrg/xdg"
+	"github.com/mitchellh/go-homedir"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+func Load(cfg Config, cmd *cobra.Command, configurations ...any) error {
+	// allow for nested options to be specified via environment variables
+	// e.g. pod.context = APPNAME_POD_CONTEXT
+	v := viper.NewWithOptions(viper.EnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_")))
+
+	return load(cfg, v, cmd, configurations...)
+}
+
+func LoadAt(cfg Config, cmd *cobra.Command, path string, configuration any) error {
+	t := reflect.TypeOf(configuration)
+	config := reflect.StructOf([]reflect.StructField{{
+		Name: upperFirst(path),
+		Type: t,
+		Tag:  reflect.StructTag(fmt.Sprintf(`json:"%s" yaml:"%s" mapstructure:"%s"`, path, path, path)),
+	}})
+
+	value := reflect.New(config)
+	value.Elem().Field(0).Set(reflect.ValueOf(configuration))
+
+	return Load(cfg, cmd, value.Interface())
+}
+
+func upperFirst(p string) string {
+	if len(p) < 2 {
+		return strings.ToUpper(p)
+	}
+	return strings.ToUpper(p[0:1]) + p[1:]
+}
+
+func load(cfg Config, v *viper.Viper, cmd *cobra.Command, configurations ...any) error {
+	for _, cfg := range configurations {
+		if reflect.TypeOf(cfg).Kind() != reflect.Ptr {
+			return fmt.Errorf("LoadConfig cfg parameter must be a pointer, got: %s -- %v", reflect.TypeOf(cfg).Name(), cfg)
+		}
+	}
+
+	// priority order: viper.Set, flag, env, config, kv, defaults
+	// flags have already been loaded into viper by command construction
+
+	// check if user specified config; otherwise read all possible paths
+	if err := loadConfig(cfg, v, cfg.AppName, cfg.ConfigFile); err != nil {
+		var notFound *viper.ConfigFileNotFoundError
+		if errors.As(err, &notFound) {
+			cfg.Log.Debug("no config file found, using defaults")
+		} else {
+			return fmt.Errorf("unable to load config: %w", err)
+		}
+	}
+
+	// load environment variables
+	v.SetEnvPrefix(cfg.AppName)
+	v.AllowEmptyEnv(true)
+	v.AutomaticEnv()
+
+	appPrefix := cfg.AppName
+	if appPrefix != "" {
+		appPrefix += "."
+	}
+
+	flags := getFlagRefs(cmd)
+
+	for _, configuration := range configurations {
+		configureViper(cfg, v, reflect.ValueOf(configuration), flags, appPrefix, "")
+
+		// unmarshal fully populated viper object onto config
+		err := v.Unmarshal(configuration)
+		if err != nil {
+			return err
+		}
+
+		// Convert all populated config options to their internal application values ex: scope string => scopeOpt source.Scope
+		err = processPostConfig(configuration)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// configureViper loads the default configuration values into the viper instance, before the config values are read and parsed
+func configureViper(cfg Config, v *viper.Viper, value reflect.Value, flags flagRefs, appPrefix string, path string) {
```

Comment:
there's lots of good tests around the effects of how this function is used on the CLI and how viper binds. I feel that this function is important enough to test with a little more depth in functionality (nested structs, odd paths, negative testing for overlapping tags [that's not compile time errors I think], etc.).

Reply:
I think having overlapping tags would be okay, maybe there's a reason for it?

Reply:
the tags was just one example, my higher order comment here is that I feel that this function needs more explicit testing of this function. The `load()` tests only look at one nested field deep, having more here would be great. Having a mix of types on the target struct, including pointers at the struct value at the end of the config path search, etc. You could add more load() tests for these cases if you want, I suggested here since all of these things seem core to this function (does the effect of the reflection and extraction get prepared into the viper object correctly).

Reply:
I've added a lot more tests for different struct fields and nesting levels.